### PR TITLE
One GPU job per GPU node

### PR DESCRIPTION
**Pull request by kathryn-zhou**
## Changes proposed in this PR

- Change the GPU constraint to only allow one GPU job on each GPU node.
- Require that the number of GPUs requested by the job is equal to the number of GPUs available on the node.

## Why are we making these changes?
- Each GPU job requires a SSD, and Google currently only allows one SSD per node.

### CODE REVIEW

File: scheduler/src/cook/kubernetes/api.clj [hunk 1]
```diff
@@ -788,7 +788,8 @@
     (when (pos? gpus)
       (.putLimitsItem resources "nvidia.com/gpu" (double->quantity gpus))
       (.putRequestsItem resources "nvidia.com/gpu" (double->quantity gpus))
-      (add-node-selector pod-spec "cloud.google.com/gke-accelerator" gpu-model-requested))
+      (add-node-selector pod-spec "cloud.google.com/gke-accelerator" gpu-model-requested)
+      (add-node-selector pod-spec "gpu-count" gpus))
```
Comment (dposada) [inline scheduler/src/cook/kubernetes/api.clj:792]: Might be nice to have an explanatory comment above this line

File: scheduler/src/cook/scheduler/constraints.clj [hunk 2]
```diff
@@ -108,14 +108,14 @@
   (job-constraint-evaluate
     [{:keys [job]} _ vm-attributes _]
     (let [k8s-vm? (= (get vm-attributes "compute-cluster-type") "kubernetes")
-          job-gpu-count-requested (-> job util/job-ent->resources :gpus (or 0))]
+          job-gpu-count-requested (-> job util/job-ent->resources :gpus (or 0) int)]
```
Comment (dposada) [inline scheduler/src/cook/scheduler/constraints.clj:111]: Might also be nice to explain why we need the `int` here

File: scheduler/src/cook/kubernetes/api.clj (hunk 1 above)
Comment (kathryn-zhou) [reply scheduler/src/cook/kubernetes/api.clj:792]: added

File: scheduler/src/cook/scheduler/constraints.clj (hunk 2 above)
Comment (kathryn-zhou) [reply scheduler/src/cook/scheduler/constraints.clj:112]: add

File: scheduler/src/cook/scheduler/constraints.clj [hunk 3]
```diff
@@ -108,14 +108,16 @@
...
           (if k8s-vm?
             (let [job-gpu-model-requested (job->gpu-model-requested
                                             job-gpu-count-requested job (util/job->pool-name job))
                   vm-gpu-model->count-available (get vm-attributes "gpus")
                   vm-satisfies-constraint? (if (pos? job-gpu-count-requested)
                                              ; If job requests GPUs, require that the VM has enough gpus available in the same model as the job requested.
-                                             (>= (get vm-gpu-model->count-available job-gpu-model-requested 0) job-gpu-count-requested)
+                                             ; cast to int because the API allows both float and int gpu values
+                                             (= (int (get vm-gpu-model->count-available job-gpu-model-requested 0)) job-gpu-count-requested)
```
Comment (nsinkov) [inline scheduler/src/cook/scheduler/constraints.clj:120]: is it better style to use (== ...) ?

**Comment by nsinkov** [comment]
Will you ever get a job that wants a number of GPUs that doesn't match the number of GPUs on any node, or is that not possible?

File: scheduler/src/cook/kubernetes/api.clj (hunk 1 above)
Comment (dposada) [reply scheduler/src/cook/kubernetes/api.clj:792]: I don't see the comment

**Comment by kathryn-zhou** [comment]
@nsinkov We are going to add in some validation for that at the job submission level. So users can only request 1, 2, 4, and 8 GPUs and if they do otherwise then the job submission will fail.

File: scheduler/src/cook/scheduler/constraints.clj (hunk 3 above)
Comment (kathryn-zhou) [reply scheduler/src/cook/scheduler/constraints.clj]: yes thank you! didn't know that existed but that is definitely cleaner and I modified this

File: scheduler/src/cook/scheduler/constraints.clj [hunk 4]
```diff
@@ -115,7 +115,7 @@
                   vm-gpu-model->count-available (get vm-attributes "gpus")
                   vm-satisfies-constraint? (if (pos? job-gpu-count-requested)
                                              ; If job requests GPUs, require that the VM has enough gpus available in the same model as the job requested.
```
Comment (dposada) [inline scheduler/src/cook/scheduler/constraints.clj:117]:
```suggestion
                                             ; If job requests GPUs, require that the VM has the same number of gpus available in the same model as the job requested.
```

**Comment by dposada** [comment]
Will merge when green

**Pull request merged**